cmd/mt-kafka-mdm-sniff-out-of-order/main.go

```diff
@@ -0,0 +1,135 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"math/rand"
+	"os"
+	"os/signal"
+	"strconv"
+	"strings"
+	"sync"
+	"syscall"
+	"text/template"
+	"time"
+
+	inKafkaMdm "github.com/raintank/metrictank/input/kafkamdm"
+	"github.com/raintank/metrictank/stats"
+	"github.com/raintank/worldping-api/pkg/log"
+	"github.com/rakyll/globalconf"
+	"gopkg.in/raintank/schema.v1"
+)
+
+var (
+	confFile = flag.String("config", "/etc/metrictank/metrictank.ini", "configuration file path")
+	format   = flag.String("format", "{{.First.Seen}} {{.First.Time}} | {{.Seen}} {{.Time}} {{.Part}} {{.OrgId}} {{.Id}} {{.Name}} {{.Metric}} {{.Interval}} {{.Value}} {{.Unit}} {{.Mtype}} {{.Tags}}", "template to render the data with")
+	prefix   = flag.String("prefix", "", "only show metrics that have this prefix")
+	substr   = flag.String("substr", "", "only show metrics that have this substring")
+
+	stdoutLock = sync.Mutex{}
+)
+
+type Data struct {
+	Part int32
+	Seen time.Time
+	schema.MetricData
+}
+
+type TplData struct {
+	Data       // currently seen
+	First Data // seen the first time
+}
+
+// find out of order metrics
+type inputOOOFinder struct {
+	template.Template
+	data map[string]Data // by metric name
+	lock sync.Mutex
+}
+
+func newInputOOOFinder(format string) *inputOOOFinder {
+	tpl := template.Must(template.New("format").Parse(format + "\n"))
+	return &inputOOOFinder{
+		*tpl,
+		make(map[string]Data),
+		sync.Mutex{},
+	}
+}
+
+func (ip *inputOOOFinder) Process(metric *schema.MetricData, partition int32) {
+	if *prefix != "" && !strings.HasPrefix(metric.Metric, *prefix) {
+		return
+	}
+	if *substr != "" && !strings.Contains(metric.Metric, *substr) {
+		return
+	}
+	now := Data{
+		Part:       partition,
+		Seen:       time.Now(),
+		MetricData: *metric,
+	}
+	ip.lock.Lock()
+	first, ok := ip.data[metric.Name]
+	if !ok {
+		ip.data[metric.Name] = now
+	} else {
+		if metric.Time > first.Time {
+			ip.data[metric.Name] = now
+		} else {
+			t := TplData{now, first}
+			stdoutLock.Lock()
```

Comment:
I dont think you need this lock, as you are already holding ip.lock.Lock() which ensures only one goroutine can call ip.Execute()